WIP feat(lldap) add lldap

MaverickD650 opened the pull request:
**Description**

Adding Lldap a lightweight alternative to OpenLDAP for those who don't need the extensive features of OpenLDAP with a user friendly interface.

⚒️ Fixes  #3753 

**⚙️ Type of change**

- [X] ⚙️ Feature/App addition
- [ ] 🪛 Bugfix
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🔃 Refactor of current code

**🧪 How Has This Been Tested?**
<!--
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
-->

**📃 Notes:**
<!-- Please enter any other relevant information here -->
Based on the enterprise Vaultwarden app to take into account new common. Still a work in progress to simplify and adapt the chart.

**✔️ Checklist:**

- [ ] ⚖️ My code follows the style guidelines of this project
- [x] 👀 I have performed a self-review of my own code
- [x] #️⃣ I have commented my code, particularly in hard-to-understand areas
- [ ] 📄 I have made corresponding changes to the documentation
- [ ] ⚠️ My changes generate no new warnings
- [ ] 🧪 I have added tests to this description that prove my fix is effective or that my feature works
- [ ] ⬆️ I increased versions for any altered app according to semantic versioning

**➕ App addition**

If this PR is an app addition please make sure you have done the following.

- [ ] 🪞 I have opened a PR on [truecharts/containers](https://github.com/truecharts/containers) adding the container to TrueCharts mirror repo.
- [ ] 🖼️ I have added an icon in the Chart's root directory called `icon.png`

---

_Please don't blindly check all the boxes. Read them and only check those that apply.
Those checkboxes are there for the reviewer to see what is this all about and
the status of this PR with a quick glance._

MaverickD650 committed:
inital

MaverickD650 committed:
Further edits

MaverickD650 committed:
Merge branch 'master' into lldap

MaverickD650 commented:
Had to go to work but leaving an initial draft for comment. Don't judge as it very definitely isn't finished or sense checked yet.

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -0,0 +1,46 @@
+                name: lldapconfig
+            - secretRef:
+                name: lldapsecret
+
```
```suggestion
configmap:
  lldapconfig:
    enabled: true
    data:
      LLDAP_BASE_DN: {{ .Values.lldap.basedn | quote }}
```

Commit:
Update charts/incubator/lldap/values.yaml

Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -0,0 +1,50 @@
+  lldapconfig:
+    enabled: true
+    data:
+      LLDAP_BASE_DN: {{ .Values.lldap.basedn | quote }}
```
```suggestion
        main:
          env:
            LLDAP_BASE_DN: {{ .Values.lldap.basedn | quote }}
          envFrom:
            - secretRef:
                name: lldapsecret
```

Commit:
Update charts/incubator/lldap/values.yaml

Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

Commit:
Delete _configmap.tpl

Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -0,0 +1,45 @@
+    ## Password for admin user
+    password: "replaceme"
+    ## User defined Base DN e.g. dc=example,dc=com
+    basedn: "dc=example,dc=com"
```
```suggestion
```

Commit:
Update charts/incubator/lldap/values.yaml

Signed-off-by: Kjeld Schouten-Lebbing <kjeld@schouten-lebbing.nl>

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -0,0 +1,43 @@
+          envFrom:
+            - secretRef:
+                name: lldapsecret
+# Set lldap variables
```
env vars should be set in the env section, unless they need to be put in a secret

Crow-Control commented on charts/incubator/lldap/templates/common.yaml:
```diff
@@ -0,0 +1,2 @@
+{{/* Render the templates */}}
+{{ include "common.all" . }}
```
secret needs to be loaded here.

Crow-Control commented on charts/incubator/lldap/templates/_secrets.tpl:
```diff
@@ -0,0 +1,21 @@
+{{- define "lldap.secrets" -}}
+
+{{- $jwtSecret := "" }}
+{{- $jwtSecret = .Values.lldap.configuration.jwt | default (randAlphaNum 48) | quote }}
```
Not going to work, this will randomise it every run.
Needs complicated lookup code.

Crow-Control commented on charts/incubator/lldap/questions.yaml:
```diff
@@ -0,0 +1,88 @@
+    schema:
+      additional_attrs: true
+      type: dict
+      attrs:
```
missing all actuall config here. Will error

Crow-Control commented on charts/incubator/lldap/questions.yaml:
```diff
@@ -0,0 +1,88 @@
+      additional_attrs: true
+      type: dict
+      attrs:
+# Include{serviceRoot}
```
Missing ldap service

MaverickD650 committed:
fixes

MaverickD650 committed:
prettier

MaverickD650 committed:
lint

MaverickD650 committed:
Oops

MaverickD650 committed:
secret name

MaverickD650 committed:
hmmm

MaverickD650 committed:
caps

MaverickD650 committed:
typo

MaverickD650 committed:
secrets

MaverickD650 committed:
more secrets

MaverickD650 committed:
yet more secrets

MaverickD650 commented:
I'm missing something with the secret and I can't seem to figure it out though definitely further along than I was. Think everything else is pretty much there so I defer to those with more knowledge on this.

Crow-Control commented on charts/incubator/lldap/templates/common.yaml:
```diff
@@ -0,0 +1,11 @@
+{{/* Render secrets for LLDAP */}}
+{{- $jwtSecret := include "lldap.secrets" . | fromYaml -}}
+{{- if $jwtSecret -}}
+  {{- $_ := set .Values.secret "lldap-secrets" $jwtSecret -}}
```
please don't duplicate the name here.
Itls "secrets" not "" lldap-secrets"

MaverickD650 committed:
set fsGroup

revert

read only

root

revert TZ

use apps user and adjust env var

missing vars

squash

hmmm

wrong chart

root and read only

prettier

hmmm

lets try this

more fixes

fixes

secret name

MaverickD650 commented on charts/incubator/lldap/values.yaml:
```diff
@@ -0,0 +1,55 @@
+    ## JWT secret set to auto generate if not defined.
+    # jwt:
+
+persistence:
+  data:
+    enabled: true
+    mountPath: "/data"
```
Currently the fsGroup, user and group is set to 0. The readOnlyRootFilesystem and runAsNonRoot is set to false. This fixed the chown with the root file system. However for this mounted volume I'm still getting "chown: /data/lldap_config.toml: Operation not permitted". I'm interpreting this as a permissions thing but can't see what else I can do to fix this.

Crow-Control replied:
We drop all capabilities, that prevents the chown.

that being said, I personally consider hard-failing chown in containers a bug.

MaverickD650 committed:
Merge branch 'master' into lldap

MaverickD650 committed:
container default user?

MaverickD650 commented:
Had a good go at this but chown and jwt secret issues suggest the container really isn't ready yet to be merged. I'll keep an eye on this and reopen if the container improves.

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -31,14 +30,14 @@ workload:
         main:
           env:
-            GID: "0"
+            GID: "1000"
```
this is automatically set

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -31,14 +30,14 @@ workload:
                   name: secrets
                   key: LLDAP_JWT_SECRET
-            UID: "0"
+            UID: "1000"
```
this is automatically set based on PUID

Crow-Control commented on charts/incubator/lldap/values.yaml:
```diff
@@ -6,12 +6,11 @@ image:
 securityContext:
   container:
     readOnlyRootFilesystem: false
-    runAsNonRoot: false
```
This is not needed in both cases, it's automatically set

jfroy commented:
I've opened a PR on lldap to fix the Dockerfile. I haven't looked at the issue with secrets (not clear what the issue is, care to provide some context/information on that?).

MaverickD650 commented:
There's a PR to auto generate the JWT token on lldap if I remember correctly

truecharts-admin commented:
This PR is locked to prevent necro-posting on closed PRs. Please create a issue or contact staff on discord if you want to further discuss this
